cis194/week10/stevemao/AParser.hs

```diff
@@ -0,0 +1,112 @@
+{-# OPTIONS_GHC -Wall #-}
+{- CIS 194 HW 10
+   due Monday, 1 April
+-}
+
+module AParser where
+
+import           Control.Applicative
+
+import           Data.Char
+
+import           Control.Monad
+
+-- A parser for a value of type a is a function which takes a String
+-- represnting the input to be parsed, and succeeds or fails; if it
+-- succeeds, it returns the parsed value along with the remainder of
+-- the input.
+newtype Parser a = Parser { runParser :: String -> Maybe (a, String) }
+
+-- For example, 'satisfy' takes a predicate on Char, and constructs a
+-- parser which succeeds only if it sees a Char that satisfies the
+-- predicate (which it then returns).  If it encounters a Char that
+-- does not satisfy the predicate (or an empty input), it fails.
+satisfy :: (Char -> Bool) -> Parser Char
+satisfy p = Parser f
+  where
+    f [] = Nothing    -- fail on the empty input
+    f (x:xs)          -- check if x satisfies the predicate
+                        -- if so, return x along with the remainder
+                        -- of the input (that is, xs)
+        | p x       = Just (x, xs)
+        | otherwise = Nothing  -- otherwise, fail
+
+-- Using satisfy, we can define the parser 'char c' which expects to
+-- see exactly the character c, and fails otherwise.
+char :: Char -> Parser Char
+char c = satisfy (== c)
+
+{- For example:
+
+*Parser> runParser (satisfy isUpper) "ABC"
+Just ('A',"BC")
+*Parser> runParser (satisfy isUpper) "abc"
+Nothing
+*Parser> runParser (char 'x') "xyz"
+Just ('x',"yz")
+
+-}
+
+-- For convenience, we've also provided a parser for positive
+-- integers.
+posInt :: Parser Integer
+posInt = Parser f
+  where
+    f xs
+      | null ns   = Nothing
+      | otherwise = Just (read ns, rest)
+      where (ns, rest) = span isDigit xs
+
+------------------------------------------------------------
+-- Your code goes below here
+------------------------------------------------------------
+
+first :: (a -> b) -> (a,c) -> (b,c)
+first f (a, c) = (f a, c)
+
+instance Functor Parser where
+  fmap f (Parser a) = Parser $ fmap (first f) . a
+
+instance Applicative Parser where
+  pure a = Parser (\s -> Just (a, s))
+  (Parser f) <*> (Parser g) = Parser fun
+    where fun a = case f a of 
+                  Nothing -> Nothing
+                  Just (fa, remaining) -> first fa <$> g remaining
+
+abParser :: Parser (Char, Char)
+abParser = (,) <$> char 'a' <*> char 'b'
+
+abParser_ :: Parser ()
+abParser_ = void abParser
+
+spaceParser :: Parser Char
+spaceParser = Parser f 
+  where f (' ' : xs) = Just (' ', xs)
+        f _ = Nothing
+
+combiner ::() -> Integer -> Char -> Integer -> [Integer]
+combiner _ s1 _ s2 = [s1,s2]
+
+sanityParser :: Parser ()
+sanityParser = Parser f
+  where f xs@[x, x', ' ', y, y'] = if isDigit x && isDigit x' && isDigit y && isDigit y'
```

Comment:
Why do we need `sanityParser`?

Reply:
The homework is a bit ambiguous. EG: it doesn't say what happens to `intPair "123 34"`

Reply:
> • Create a parser intPair which reads two integer values separated
> by a space and returns the integer values in a list. You should use
> the provided posInt to parse the integer values.
In the case of "123 34", both "123" and "34" are integer values.